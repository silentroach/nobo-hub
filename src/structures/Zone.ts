import { escapeSpace, unescapeSpace } from '../utils';

export interface Zone {
	/**
	 * Ignored when you add a zone, will return new id
	 */
	id: number;
	/**
	 * Name of zone. Should be no longer than 100 bytes
	 * Note that spaces will be encoded as 2 bytes
	 */
	name: string;
	/**
	 * Optional ID of Active Week Profile. Default week profile will be used If not specified
	 */
	activeWeekProfileId?: number;
	/**
	 * Degrees Celsius. Expected to be an integer at the range 7 to 30
	 * and at least as large as the Eco temperature
	 */
	comfortTemperature: number;
	/**
	 * Degrees Celsius. Expected to be an integer at the range 7 to 30
	 * and no larger than the Comfort temperature
	 */
	ecoTemperature: number;
	allowOverrides: boolean;
}

export const validate = (zone: Zone) => {
	if (zone.name === '') {
		throw new TypeError('Empty zone name is not allowed');
	}

	if (Buffer.byteLength(escapeSpace(zone.name)) > 100) {
		throw new TypeError(
			'Zone name is too long, maximum 100 bytes allowed (note that spaces are encoded as 2 bytes)'
		);
	}

	if (zone.activeWeekProfileId && zone.activeWeekProfileId < 1) {
		throw new TypeError('Invalid zone active week profile id');
	}

	if (zone.comfortTemperature < 7 || zone.comfortTemperature > 30) {
		throw new TypeError(
			'Invalid comfort temperature, should be between 7 and 30 celsius'
		);
	}

	if (zone.ecoTemperature < 7 || zone.ecoTemperature > 30) {
		throw new TypeError(
			'Invalid eco temperature, should be between 7 and 30 celsius'
		);
	}

	if (zone.ecoTemperature > zone.comfortTemperature) {
		throw new TypeError(
			'Invalid temperature, eco temperature should be less than comfort temperature'
		);
	}
};

// <Zone id> <Name> <Active week profile id> <Comfort temperature> <Eco temperature> <Allow overrides> <Active override id>
const ZoneRegexp = /^(?<id>\d+)\s(?<name>.*?)\s(?<week>\d+)\s(?<comfort>\d+)\s(?<eco>\d+)\s(?<overrides>[10])\s-1$/;

export const deserialize = (input: string): Zone => {
	const matches = input.match(ZoneRegexp);
	if (matches === null) {
		throw new SyntaxError('Invalid zone info structure');
	}

	const groups = matches.groups as {
		id: string;
		name: string;
		week: string;
		comfort: string;
		eco: string;
		overrides: string;
	};

	return {
		id: Number(groups.id),
		name: unescapeSpace(groups.name),
		activeWeekProfileId: Number(groups.week),
		comfortTemperature: Number(groups.comfort),
		ecoTemperature: Number(groups.eco),
		allowOverrides: Boolean(Number(groups.overrides)),
	};
};

export const serialize = (zone: Zone) =>
	[
		zone.id,
		escapeSpace(zone.name),
		// 1 is readonly default week profile id
		zone.activeWeekProfileId !== undefined ? zone.activeWeekProfileId : 1,
		zone.comfortTemperature,
		zone.ecoTemperature,
		zone.allowOverrides ? 1 : 0,
		/** active override id, reserved, not used */ -1,
	].join(' ');
