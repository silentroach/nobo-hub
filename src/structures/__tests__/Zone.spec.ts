import { deserialize, serialize, validate, Zone } from '../Zone';

describe('validate', () => {
	it('validate minimal valid info', () => {
		validate({
			id: 1,
			name: 'zone',
			comfortTemperature: 7,
			ecoTemperature: 7,
			allowOverrides: false,
		});
	});

	it('validate throws on empty name', () => {
		expect(() =>
			validate({
				id: 1,
				name: '',
				comfortTemperature: 7,
				ecoTemperature: 7,
				allowOverrides: false,
			})
		).toThrow();
	});

	it('validate throws on name is too long', () => {
		expect(() =>
			validate({
				id: 1,
				name: ['name is just t', 'o'.repeat(100), ' long'].join(' '),
				comfortTemperature: 7,
				ecoTemperature: 7,
				allowOverrides: false,
			})
		).toThrow();
	});

	it('validate throws on invalid temperature', () => {
		expect(() =>
			validate({
				id: 1,
				name: 'zone',
				comfortTemperature: -5,
				ecoTemperature: 7,
				allowOverrides: false,
			})
		).toThrow();

		expect(() =>
			validate({
				id: 1,
				name: 'zone',
				comfortTemperature: 5,
				ecoTemperature: 7,
				allowOverrides: false,
			})
		).toThrow();

		expect(() =>
			validate({
				id: 1,
				name: 'zone',
				comfortTemperature: 35,
				ecoTemperature: 7,
				allowOverrides: false,
			})
		).toThrow();

		expect(() =>
			validate({
				id: 1,
				name: 'zone',
				comfortTemperature: 9,
				ecoTemperature: -5,
				allowOverrides: false,
			})
		).toThrow();

		expect(() =>
			validate({
				id: 1,
				name: 'zone',
				comfortTemperature: 9,
				ecoTemperature: 4,
				allowOverrides: false,
			})
		).toThrow();

		expect(() =>
			validate({
				id: 1,
				name: 'zone',
				comfortTemperature: 9,
				ecoTemperature: 35,
				allowOverrides: false,
			})
		).toThrow();
	});

	it('validate throws on eco temperature is greater than comfort temperature', () => {
		expect(() =>
			validate({
				id: 1,
				name: 'zone',
				comfortTemperature: 10,
				ecoTemperature: 15,
				allowOverrides: false,
			})
		).toThrow();
	});

	it('validate throws on invalid week profile id', () => {
		expect(() =>
			validate({
				id: 1,
				name: 'name',
				activeWeekProfileId: -5,
				comfortTemperature: 7,
				ecoTemperature: 7,
				allowOverrides: false,
			})
		).toThrow();
	});
});

describe('serialization', () => {
	it('should serialize/deserialize minimal component info', () => {
		const zone: Zone = {
			id: 1,
			name: 'zone',
			comfortTemperature: 7,
			ecoTemperature: 7,
			allowOverrides: false,
		};

		const serialized = serialize(zone);
		expect(serialized).toMatchSnapshot();

		const deserialized = deserialize(serialized);

		zone.activeWeekProfileId = 1; // by default
		expect(zone).toEqual(deserialized);
	});

	it('should escape spaces in name', () => {
		const zone: Zone = {
			id: 1,
			name: 'zone with spaces',
			comfortTemperature: 7,
			ecoTemperature: 7,
			allowOverrides: false,
		};

		const serialized = serialize(zone);
		expect(serialized).toMatchSnapshot();

		const deserialized = deserialize(serialized);

		zone.activeWeekProfileId = 1; // by default
		expect(zone).toEqual(deserialized);
	});

	it('name with utf', () => {
		const zone: Zone = {
			id: 1,
			name: 'zone ðŸ˜Ž',
			comfortTemperature: 7,
			ecoTemperature: 7,
			allowOverrides: false,
		};

		const serialized = serialize(zone);
		expect(serialized).toMatchSnapshot();

		const deserialized = deserialize(serialized);

		zone.activeWeekProfileId = 1; // by default
		expect(zone).toEqual(deserialized);
	});
});
