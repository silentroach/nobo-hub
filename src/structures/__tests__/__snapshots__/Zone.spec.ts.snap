// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serialization name with utf 1`] = `"1 zoneÂ ðŸ˜Ž 1 7 7 0 -1"`;

exports[`serialization should escape spaces in name 1`] = `"1 zoneÂ withÂ spaces 1 7 7 0 -1"`;

exports[`serialization should serialize/deserialize minimal component info 1`] = `"1 zone 1 7 7 0 -1"`;
